BuildKit and build-arg support

craig-willis opened the pull request:
This PR adds the following:
* `build-arg` flag support:  buildpacks can specify `ARG` commands and the `repo2docker` cli supports `--build-arg` flags to set these values at build time
* Docker [build-kit](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md) support including moving from `docker-py` to Docker CLI for build operations

See also:
*  https://github.com/docker/docker-py/issues/2230
* https://github.com/docker/compose/pull/6865/files

craig-willis committed:
Added build-arg support

craig-willis committed:
Adds buildkit support

Xarthisius commented on repo2docker/__main__.py:
```diff
@@ -359,6 +367,11 @@ def make_r2d(argv=None):
 
+    r2d.extra_build_args = {}
+    for build_arg in args.extra_build_args:
+        kv = build_arg.split(":")
```
Docstring says `key=value` yet you're splitting it using `:` ?

Xarthisius commented on repo2docker/docker_utils.py:
```diff
@@ -0,0 +1,63 @@
+
+        rc = p.poll()
+        if rc != 0:
+          yield {"error": line}
```
not multiple of 4 whitespaces. Instead of fixing this manually could you run this file through `black` ?

Xarthisius commented on repo2docker/app.py:
```diff
@@ -128,6 +128,16 @@ def _default_log_level(self):
+        {},
+        help="""
+        Regular Docker build-time  arguments that will be passed as
+        --build-arg key=value.
```
```suggestion
        --build-arg key:value.
```

craig-willis committed:
Reformatted with black

craig-willis committed:
Split on =

craig-willis committed:
Fixed rebase mistake

craig-willis committed:
Fixed misuse of process.poll()

craig-willis commented:
Rebased and fixed the problem I was seeing with how I was using `poll`.  Popen.poll()` returns `None` when the child process hasn't terminated. I was checking for an empty string from readline(), which caused problems.  Also, I ran black and fixed a problem but the CI linting is failing and I don't know why.

Xarthisius committed:
Add missing comma to make black happy

craig-willis committed:
Changed to Config/WorkingDir

craig-willis committed:
Convert docker CLI build to class

craig-willis committed:
Path handling for DockerBuildpack

craig-willis committed:
Change tests to use DockerCLI

craig-willis committed:
Add test for build args

craig-willis committed:
Only set mran date in detect

craig-willis committed:
Fix r test for now

Xarthisius committed:
Bring back autodection of latest mran
